# Homework - 2
***
**Name**: Shipra Behera
***

This assignment is due on Canvas by **5pm on Friday October 5th**. Submit only this Jupyter notebook to Canvas.  Do not compress it using tar, rar, zip, etc. Your solutions to analysis questions should be done in Markdown directly below the associated question.  Remember that you are encouraged to discuss the problems with your classmates and instructors, but **you must write all code and solutions on your own**, and list any people or sources consulted.

## Boosting - Extra Credit [5-points]
***

In this problem, we slightly modify the AdaBoost algorithm to better explore some properties of the algorithm. Specifically, we no longer normalize the weights on the training examples after each iteration. The modified algorithm, which is set to run for $T$ iterations, is shown in Algorithm I.

Note that in the modified version, the weights associated with the training examples are no longer guaranteed to sum to one after each iteration (and therefore can not be viewed as a "distribution"), but the algorithm is still valid. Let us denote the sum of weights at the start of iteration $t$ by $Z_t = \sum_{i=1}^{n}w_i^{(t)}$. At the start of the first iteration of boosting, $Z_1 = n$. Let us now investigate the behavior of $Z_t$, as a function of t

![image](fig-1.png)

**A:** At the $t^{th}$ iteration, we found a weak classifier that achieves a weighted training error $\epsilon_t$. Show that the choice, $\alpha_t = \frac{1}{2}\log\frac{1 - \epsilon_t}{\epsilon_t}$ is the optimal in the sense that it minimizes $Z_{t+1}$

*Hint: Look at $Z_{t+1}$ as function of $\alpha$ and find the value of $\alpha$ for which the function achieves the minimum. You may also find the following notational shorthand useful:

$$W_t = \sum_{i=1}^{n}w_i^{(t)}(1 - \delta(y_i, h_t(x_i)))$$
$$W_c = \sum_{i=1}^{n}w_i^{(t)}(\delta(y_i, h_t(x_i)))$$

where $W_c$ is the total weight of the points classified correctly by $h_t$ and $W_t$ is the total weight of the misclassified points. $\delta(y, h_t(x)) = 1$ whenever the label predicted by $h_t$ is correct and zero otherwise. The weights here are those available at the start of iteration $t$

**Solution:** <br/>
$y_i$ is the true label<br/>
$h_t(x_i)$ is the output of the weak classifier $h_t$<br/>
Since, it is a binary classification problem, for misclassified points, $y_ih_t(x_i)$ would be -1. And for correctly classified points, $y_ih_t(x_i)$ would be +1.<br/><br/>
Updated weight for misclassified points, $w_i^{(t+1)} = e^{\alpha_t}$  <br/> 
Total weight for misclassified points at $t^{th}$ iteration, <br/>
$W_t = \sum_{i=1}^{n}w_i^{(t)}(1 - \delta(y_i, h_t(x_i)))$ <br/>
$\implies W_t = \sum_{i=1}^{n}w_i^{(t)}$, since $\delta(y, h_t(x)) = 0$ <br/>
Total weight for misclassified points at $t+1^{th}$ iteration, <br/>
$W_{t+1} = \sum_{i=1}^{n}w_i^{(t+1)}$ <br/>
$\implies W_{t+1} = \sum_{i=1}^{n}w_i^{(t)}e^{\alpha_t}$ <br/>
$\implies W_{t+1} = W_te^{\alpha_t}$ <br/><br/>
Updated weight for correctly classified points, $w_i^{(t+1)} = e^{-\alpha_t}$ <br/>
Similarly, Total weight for correctly classified points at $t+1^{th}$ iteration, <br/>
$W_{c+1} = W_ce^{-\alpha_t}$ <br/><br/>
Sum of weights at $t+1^{th}$ iteration as a function of $\alpha$, <br/>
$Z_{t+1}(\alpha_t) = W_{t+1} + W_{c+1}$ <br/>
$\implies Z_{t+1}(\alpha_t) = W_te^{\alpha_t} + W_ce^{-\alpha_t}$ 

From the definition of weighted training error, we see that <br/>
$\epsilon_t = \frac{1}{Z_t}\sum_{i=1}^{n}w_i^{(t)}(1 - \delta(y_i, h_t(x_i)))$ <br/>
$\implies \epsilon_t = \frac{W_t}{Z_t}$ <br/>
$\implies W_t = Z_t\epsilon_t$ <br/>
Similary, $W_c = Z_t{(1-\epsilon_t)}$

To find the value for which $Z_{t+1}(\alpha_t)$ achieves the minimum, we take its derivative w.r.t $\alpha_t$ where it is 0.<br/>
$\frac{\partial Z_{t+1}}{\partial \alpha_t} = W_te^{\alpha_t} - W_ce^{-\alpha_t} = 0$<br/>
$\implies e^{2\alpha_t} = \frac{W_c}{W_t}$ <br/>
$\implies \alpha_t = \frac{1}{2}\log\frac{W_c}{W_t}$ <br/>

From the values of $W_t$ and $W_c$ from above, it is proved that <br/>
$\alpha_t = \frac{1}{2}\log\frac{1 - \epsilon_t}{\epsilon_t}$ is optimal.

**B:** Show that the sum of weights $Z_t$ is monotonically decreasing as a function of $t$.

**Solution:** <br/>
$Z_{t+1}(\alpha_t) = W_te^{\alpha_t} + W_ce^{-\alpha_t}$ 
At $\alpha_t = 0$, <br/>
$Z_{t+1}(0) = W_t + W_c = Z_t$<br/> 
Hence, $Z_{t+1} = \min_{\alpha_t} Z_{t+1} (\alpha_t) < Z_{t+1} (0) = Z_t$, when the minimizing $\alpha_t$ is non-zero. <br/>
$\alpha_t = \frac{1}{2}\log\frac{1 - \epsilon_t}{\epsilon_t}$ is non-zero when <br/>
$\log\frac{1 - \epsilon_t}{\epsilon_t} = 0 $<br/>
$\implies \frac{1 - \epsilon_t}{\epsilon_t} = 1 $<br/>
$\implies {\epsilon_t} = \frac{1}{2} $ i.e when error of the weak classifier is $\frac{1}{2}$ <br/>

To see how much $Z_t$ decreases as a function of t, <br/>
$Z_{t+1} = W_te^{-\frac{1}{2}\log\frac{W_c}{W_t}} + W_ce^{\frac{1}{2}\log\frac{W_c}{W_t}}$ <br/>
= $2\sqrt {W_tW_c}$ <br/>

$Z_t = W_t + W_c$, where $W_t < W_c$ <br/>
We can check if $Z_{t+1}$ is less than $Z_t$ <br/>
$Z_t^2 - Z_{t+1}^2 = W_t^2 + W_c^2 + 2W_tW_c - 4W_tW_c$ <br/>
= $W_t^2 + W_c^2 - 2W_tW_c$ <br/>
= $(W_c - W_t)^2 $ which is > 0, <br/>
which proves that $Z_t$ is monotonically decreasing as a function of $t$ .

## Training Data
***
Please do not change this class

In [6]:
import numpy as np
from sklearn.base import clone

In [7]:
class ThreesAndEights:
    """
    Class to store MNIST data
    """

    def __init__(self, location):

        import pickle, gzip

        # Load the dataset
        f = gzip.open(location, 'rb')

        # Split the data set 
#         X_train, y_train, X_valid, y_valid = pickle.load(f)
        train_set, valid_set, test_set = pickle.load(f)
    
        X_train, y_train = train_set
        X_valid, y_valid = valid_set

        # Extract only 3's and 8's for training set 
        self.X_train = X_train[np.logical_or( y_train==3, y_train == 8), :]
        self.y_train = y_train[np.logical_or( y_train==3, y_train == 8)]
        self.y_train = np.array([1 if y == 8 else -1 for y in self.y_train])
        
        # Shuffle the training data 
        shuff = np.arange(self.X_train.shape[0])
        np.random.shuffle(shuff)
        self.X_train = self.X_train[shuff,:]
        self.y_train = self.y_train[shuff]
        print("Training set ", len(self.X_train))
        # Extract only 3's and 8's for validation set 
        self.X_valid = X_valid[np.logical_or( y_valid==3, y_valid == 8), :]
        print("Validation set ", len(self.X_valid))
        self.y_valid = y_valid[np.logical_or( y_valid==3, y_valid == 8)]
        self.y_valid = np.array([1 if y == 8 else -1 for y in self.y_valid])
        f.close()

In [8]:
data = ThreesAndEights("data/mnist.pklz")
print(len(data.X_valid))

Training set  9943
Validation set  2039
2039


Feel free to explore this data and get comfortable with it before proceeding further.

## Bagging
Bootstrap aggregating, also called bagging, is a machine learning ensemble meta-algorithm designed to improve the stability and accuracy of machine learning algorithms used in statistical classification and regression. It also reduces variance and helps to avoid overfitting. Although it is usually applied to decision tree methods, it can be used with any type of method. Bagging is a special case of the model averaging approach.

Given a standard training set $D$ of size n, bagging generates $N$ new training sets $D_i$, roughly each of size n * ratio, by sampling from $D$ uniformly and with replacement. By sampling with replacement, some observations may be repeated in each $D_i$ The $N$ models are fitted using the above $N$ bootstraped samples and combined by averaging the output (for regression) or voting (for classification). 

-Source [Wiki](https://en.wikipedia.org/wiki/Bootstrap_aggregating)

## Implementing Bagging [5-points]
***

We've given you a skeleton of the class `BaggingClassifier` below which will train a classifier based on the decision trees as implemented by sklearn. Your tasks are as follows, please approach step by step to understand the code flow:
* Implement `bootstrap` method which takes in two parameters (`X_train, y_train`) and returns a bootstrapped training set ($D_i$)
* Implement `fit` method which takes in two parameters (`X_train, y_train`) and trains `N` number of base models on different bootstrap samples. You should call `bootstrap` method to get bootstrapped training data for each of your base model
* Implement `voting` method which takes the predictions from learner trained on bootstrapped data points `y_hats` and returns final prediction as per majority rule. In case of ties, return either of the class randomly.
* Implement `predict` method which takes in multiple data points and returns final prediction for each one of those. Please use the `voting` method to reach consensus on final prediction.

In [280]:
from sklearn.tree import DecisionTreeClassifier

class BaggingClassifier:
    def __init__(self, ratio = 0.20, N = 20, base=DecisionTreeClassifier(max_depth=4)):
        """
        Create a new BaggingClassifier
        
        Args:
            base (BaseEstimator, optional): Sklearn implementation of decision tree
            ratio: ratio of number of data points in subsampled data to the actual training data
            N: number of base estimator in the ensemble
        
        Attributes:
            base (estimator): Sklearn implementation of decision tree
            N: Number of decision trees
            learners: List of models trained on bootstrapped data sample
        """
        
        assert ratio <= 1.0, "Cannot have ratio greater than one"
        self.base = base
        self.ratio = ratio
        self.N = N
        self.learners = []
        
    def fit(self, X_train, y_train):
        """
        Train Bagging Ensemble Classifier on data
        
        Args:
            X_train (ndarray): [n_samples x n_features] ndarray of training data   
            y_train (ndarray): [n_samples] ndarray of data 
        """
        #TODO: Implement functionality to fit models on the bootstrapped samples
        # cloning sklearn models:
        # from sklearn.base import clone
        # h = clone(self.base)
        from sklearn.base import clone
        for i in range(0, self.N):
            Dx, Dy = self.boostrap(X_train, y_train)
            h = clone(self.base)
            self.learners.append(h.fit(Dx, Dy))
            '''test = h.predict(Dx)
            print(test)
            print(Dy)'''
        
        
    def boostrap(self, X_train, y_train):
        """
        Args:
            n (int): total size of the training data
            X_train (ndarray): [n_samples x n_features] ndarray of training data   
            y_train (ndarray): [n_samples] ndarray of data 
        """
        import random
        #get all available indices in an array
        idx = [i for i in range(len(X_train))]
        #generate N training sets of size len(training_set)*ratio, sample from D uniformly with replacement
        Dx = []
        Dy = []
        #select s number of indices from idx array
        size = int(float(len(X_train)) * float(self.ratio))
        samples = random.choices(idx, k=size)
        Dx = X_train[samples]
        Dy = y_train[samples]
      
        return Dx, Dy
        
    
    def predict(self, X):
        """
        BaggingClassifier prediction for data points in X
        
        Args:
            X (ndarray): [n_samples x n_features] ndarray of data 
            
        Returns:
            yhat (ndarray): [n_samples] ndarray of predicted labels {-1,1}
        """
                
        yhat = []
        for i in range(len(X)):
            pred = []
            for j in range(len(self.learners)):
                output = self.learners[j].predict(X)[i]
                pred.append(output)
            yhat.append(self.voting(pred))
        return yhat
        #TODO: Using the individual classifiers trained predict the final prediction using voting mechanism
    
    def voting(self, y_hats):
        """
        Args:
            y_hats (ndarray): [N] ndarray of data
        Returns:
            y_final : int, final prediction of the 
        """
        #TODO: Implement majority voting scheme and incase of ties return random label
        from collections import Counter
        labels = Counter(y_hats)
        y_final = labels.most_common(1)[0][0]
        return y_final
        

## BaggingClassifier for Handwritten Digit Recognition [5-points]
***

After you've successfully completed `BaggingClassifier` find the optimal values of `ratio`, `N` and `depth` using k-fold cross validation. You are allowed to use sklearn library to split your training data in folds. Use the data from `ThreesAndEights` class initialized variable `data`. 

Justify why those values are optimal.

Report accuracy on the validation data using the optimal parameter values.

In [1378]:
import random
bagging = BaggingClassifier(N = 1)
#print(len(data.X_train))
'''bagging.fit(data.X_valid, data.y_valid)
y_predict = bagging.predict(data.X_valid)
accuracy = 0
for i in range(len(y_predict)):
    if (y_predict[i] == data.y_valid[i]):
        accuracy = accuracy + 1
print("Accuracy is", accuracy*100/len(y_predict))'''
from sklearn.model_selection import KFold
kf = KFold(n_splits=3)

params = []
for i in range(1, 6):
    param = []
    #ratio
    param.append(round(0.1 * (i+2), 1))
    #N
    param.append(i*5)
    #depth
    param.append(i*5)
    params.append(param)
print(params)    
itr =0
idx = 0
final_acc = 0
for itr in range(len(params)):
    print("Iteration ", itr)
    print("Ratio: ", params[itr][0], ", N: ", params[itr][1], ", Max Depth: ", params[itr][2])
    bagging = BaggingClassifier(ratio = params[itr][0], N = params[itr][1], base=DecisionTreeClassifier(max_depth=params[itr][2]) )       
    accu_mean = 0
    accu_t = 0
    for train_index, test_index in kf.split(data.X_train):
        X_train, X_test = data.X_train[train_index], data.X_train[test_index]
        y_train, y_test = data.y_train[train_index], data.y_train[test_index]
        
        bagging.fit(X_train, y_train)
        y_predict = bagging.predict(X_test)
        accuracy = 0
        for i in range(len(y_predict)):
            if (y_predict[i] == y_test[i]):
                accuracy = accuracy + 1
        accu_t  += accuracy*100/len(y_predict)
    accu_mean = accu_t/3
    print("Accuracy for this iteration", accu_mean)
    if  accu_mean > final_acc:
        final_acc = accu_mean
        idx = itr
        #itr += 1


[[0.3, 5, 5], [0.4, 10, 10], [0.5, 15, 15], [0.6, 20, 20], [0.7, 25, 25]]
Iteration  0
Ratio:  0.3 , N:  5 , Max Depth:  5
Accuracy for this iteration 96.22854183221965
Iteration  1
Ratio:  0.4 , N:  10 , Max Depth:  10
Accuracy for this iteration 97.6768060178902
Iteration  2
Ratio:  0.5 , N:  15 , Max Depth:  15
Accuracy for this iteration 98.11938200840895
Iteration  3
Ratio:  0.6 , N:  20 , Max Depth:  20
Accuracy for this iteration 98.46130786313256
Iteration  4
Ratio:  0.7 , N:  25 , Max Depth:  25
Accuracy for this iteration 98.57197689282606


**Solution:** <br/>
From the k-fold cross validation, although we see that the maximum accuracy is when Ratio = 0.7 , N = 25 , Max Depth = 25. <br/>
This certain permutation of the parameters  might not be optimal because we can clearly see that accuracy does not increase significantly beyond a certain depth (10 to 15). Increasing the depth would only overfit the data, hence it is a bad idea to do so. <br/>
We can however claim that the increasing accuracies are a result of increasing the number of learners (N=25). Increasing the ratio also has some impact on the overall accuracy as higher sample size, makes the classifiers very generalizable. <br/>
Hence, we can conclude that optimal params lie somewhere between Ratio:  0.7 , N:  25 , Max Depth:  15 <br/>
Accuracy on these parameters is reported as below.

In [1383]:

#find accuracy on validation set
bagging = BaggingClassifier(ratio = 0.7, N = 25, base=DecisionTreeClassifier(max_depth=15) )
bagging.fit(X_train, y_train)
y_predict = bagging.predict(data.X_valid)
accuracy = 0
for i in range(len(y_predict)):
    if (y_predict[i] == data.y_valid[i]):
        accuracy = accuracy + 1
print("Accuracy on validation set", accuracy*100/len(y_predict))    

Accuracy on validation set 97.00833742030407


# Random Decision Tree [10-points]

In this assignment you are going to implement a random decision tree using random vector method as discussed in the lecture.

Best split: One that achieves maximum reduction in gini index across multiple candidate splits. (decided by `candidate_splits` attribute of the class `RandomDecisionTree`)

Use `TreeNode` class as node abstraction to build the tree

You are allowed to add new attributes in the `TreeNode` and `RandomDecisionTree` class - if that helps.

Your tasks are as follows:
* Implement `gini_index` method which takes in class labels as parameter and returns the gini impurity as measure of uncertainty

* Implement `majority` method which picks the most frequent class label. In case of tie return any random class label

* Implement `find_best_split` method which finds the random vector/hyperplane which causes most reduction in the gini index. 

* Implement `build_tree` method which uses `find_best_split` method to get the best random split vector for current set of training points. This vector partitions the training points into two sets, and you should call `build_tree` method on two partitioned sets and build left subtree and right subtree. Use `TreeNode` as abstraction for a node.

> The method calls itself recursively to the generate left and right subtree till the point either `max_depth` is reached or no good random split is found.  When either of two cases is encountered, you should make that node as leaf and identify the label for that leaf to be the most frequent class (use `majority` method). Go through lecture slides for better understanding

* Implement `predict` method which takes in multiple data points and returns final prediction for each one of those using the tree built. (`root` attribute of the class)

In [9]:
class TreeNode:
    def __init__(self):
        self.left = None
        self.right = None
        self.isLeaf = False
        self.label = None
        self.split_vector = None
        self.features = None
        self.labels = None

    def getLabel(self):
        if not self.isLeaf:
            raise Exception("Should not to do getLabel on a non-leaf node")
        return self.label
    
class RandomDecisionTree:
            
    def __init__(self, candidate_splits = 100, depth = 10):
        """
        Args:
            candidate_splits (int) : number of random decision splits to test
            depth (int) : maximum depth of the random decision tree
        """
        self.candidate_splits = candidate_splits
        self.depth = depth
        self.root = None
        self.gini_red = 0
        
    def fit(self, X_train, y_train):
        """
        Args:
            X_train (ndarray): [n_samples x n_features] ndarray of training data   
            y_train (ndarray): [n_samples] ndarray of data
            
        """
        self.root = self.build_tree(X_train[:], y_train[:], 0)
        return self
        
    def terminal_node(self, labels):
        return self.majority(labels)

        
    def build_tree(self, X_train, y_train, height):
        """
        Args:
            X_train (ndarray): [n_samples x n_features] ndarray of training data   
            y_train (ndarray): [n_samples] ndarray of data
            
        """
        
        if len(X_train) == 0:
            node = TreeNode()
            node.isLeaf = True
            node.labels = self.majority(y_train)
            node.features = X_train
            return node
        if height >= self.depth:
            node = TreeNode()
            node.isLeaf = True
            node.labels = self.majority(y_train)
            node.features = X_train
            return node
        else:
            node = TreeNode()
            
            split_vector = self.find_best_split(X_train, y_train)
            if split_vector is None:
                node.isLeaf = True
                node.features = X_train
                node.labels = self.majority(y_train)
                return node

            node.split_vector = split_vector
            
            leftarr, rightarr, leftlabel, rightlabel = self.split(node.split_vector, X_train, y_train)
            node.left = self.build_tree(leftarr, leftlabel, height+1)

            node.right = self.build_tree(rightarr, rightlabel, height+1)
            return node
    

    # Split a dataset based on an attribute and an attribute value
    def split(self, vector, dataset, labels):
        left, right = [], []
        leftlabel, rightlabel = list(), list()
        for i in range(0, len(dataset)):
            dp = np.dot(dataset[i], np.transpose(vector))
            if dp < 0:
                left.append(dataset[i])
                leftlabel.append(labels[i])
            else:
                right.append(dataset[i])
                rightlabel.append(labels[i])
        return np.array(left), np.array(right), leftlabel, rightlabel
    
    def find_best_split(self, X_train, y_train):
        """
        Args:
            X_train (ndarray): [n_samples x n_features] ndarray of training data   
            y_train (ndarray): [n_samples] ndarray of data
            
        """
        split_vector = None
        # your logic here
        
        import numpy as np
        
        #initialize random vector of len 784
        import random
        #split_vector = np.random.randn(1, len(X_train[0]))
        final_vector = None
        #print(len(X_train[0]))
        #get gini index of whole set
        gi_d = self.gini_index(y_train)
        cs = self.candidate_splits
        gi_min = 999
        while cs != 0:
            split_vector = np.random.normal(0, 1, len(X_train[0]))
            left, right, leftlabel, rightlabel = self.split(split_vector, X_train, y_train)
            #get gini index of left group 
            gi_l = self.gini_index(leftlabel)
            #get gini index of right group
            gi_r = self.gini_index(rightlabel)
            #weighted gini 
            gini = ((len(leftlabel)/len(y_train)) * gi_l) + ((len(rightlabel)/len(y_train)) * gi_r)
            if (gini < gi_d and gini < gi_min and gini > 0):
                gi_min = gini
                final_vector = split_vector
            cs -= 1
        return final_vector
            
        
    def gini_index(self, y):
        """
        Args:
            y (ndarray): [n_samples] ndarray of data
        """
        import collections
        import math
        countdict = collections.Counter(y)
        prob = 0
        for c in countdict.keys():
            prob += math.pow((countdict[c]/len(y)), 2)
        
        gi = 1 - prob
        return gi
    
    def majority(self, y):
        """
        Return the major class in ndarray y
        """
        from collections import Counter
        labels = Counter(y)
        y_final = labels.most_common(1)[0][0]
        return y_final
    
    def predict(self, tree, X):
        """
        BaggingClassifier prediction for new data points in X
        
        Args:
            X (ndarray): [n_samples x n_features] ndarray of data 
            
        Returns:
            yhat (ndarray): [n_samples] ndarray of predicted labels {-1,1}
        """
        if tree.isLeaf == True or tree.split_vector is None:
            return tree.labels
        dp = np.dot(X, np.transpose(tree.split_vector))
        if dp < 0:
            if tree.left != None:
                return self.predict(tree.left, X)
            else:
                print("Left not found")
                return tree.labels
        else:
            if tree.right != None:
                return self.predict(tree.right, X)
            else:
                print("Right not found")
                return tree.labels

## RandomDecisionTree for Handwritten Digit Recognition

After you've successfully completed `RandomDecisionTree`, and train using the default values in the constructor and report accuracy on the test set. Use the data from `ThreesAndEights` class initialized variable `data` 

In [10]:
dt = RandomDecisionTree()
tree = dt.fit(data.X_train, data.y_train)
predictions = []
for row in data.X_valid:
    prediction = dt.predict(tree.root, row)
    predictions.append(prediction)
accuracy = 0
for i in range(len(data.y_valid)):
    if (predictions[i] == data.y_valid[i]):
        accuracy = accuracy + 1
print("Accuracy on validation set", accuracy*100/len(predictions))  

Accuracy on validation set 91.0740559097597


# Random Forest [5-points]
Random forests or random decision forests are an ensemble learning method for classification, regression and other tasks, that operate by constructing a multitude of decision trees at training time and outputting the class that is the mode of the classes (classification) or mean prediction (regression) of the individual trees. Random decision forests correct for decision trees' habit of overfitting to their training set.

Random forest trains random decision trees on bootstrapped training points. Thus, you can implementation of methods (`bootstrap`, `predict`) from `BaggingClassifier` class directly. Only difference being, you have to use the `RandomDecisionTree` as base which you implemented previously instead of sklearn's implementation of `DecisionTreeClassifier`). Implement the `fit` method in the class below accordingly.

In [1298]:
class RandomForest(BaggingClassifier):
    def __init__(self, ratio = 0.20, N = 20, max_depth = 10, candidate_splits = 100):
        self.ratio = ratio
        self.N = N
        self.learners = []
        self.candidate_splits = candidate_splits
        self.max_depth = max_depth
        self.dt = RandomDecisionTree(candidate_splits = candidate_splits, depth = max_depth)   
    def fit(self, X_train, y_train):
        """
        Train Bagging Ensemble Classifier on data
        
        Args:
            X_train (ndarray): [n_samples x n_features] ndarray of training data   
            y_train (ndarray): [n_samples] ndarray of data 
        """
        for i in range(0, self.N):
            Dx, Dy = BaggingClassifier.boostrap(self, X_train, y_train)
            tree = self.dt.fit(Dx, Dy)
            self.learners.append(tree)
    
    def predict(self, X):
        """
        BaggingClassifier prediction for data points in X
        
        Args:
            X (ndarray): [n_samples x n_features] ndarray of data 
            
        Returns:
            yhat (ndarray): [n_samples] ndarray of predicted labels {-1,1}
        """
                
        yhat = []
        for row in X:
            pred = []
            for j in range(len(self.learners)):
                output = dt.predict(self.learners[j].root, row)
                pred.append(output)
            yhat.append(BaggingClassifier.voting(self, y_hats = pred))
        return yhat
        #TODO: Using the individual classifiers trained predict the final prediction using voting mechanism
          
        

## RandomForest for Handwritten Digit Recognition [5-points]
***

After you've successfully completed `RandomForest` find the optimal values of `ratio`, `N`, `candidate_splits` and `depth` using k-fold cross validation on. Feel free to use sklearn library to split your training data. Use the data from `ThreesAndEights` class intialized variable `data`. 

Justify why those values are optimal.

Report best accuracy on the testing data using those optimal parameter values.

In [1369]:
params = []
for i in range(1, 4):
    param = []
    #ratio
    param.append(round(0.15 * (i+2), 1))
    #N
    param.append(i*10)
    #depth
    param.append(i*5)
    #candidate splits
    param.append(i*50)
    params.append(param)
print(params) 

[[0.4, 10, 5, 50], [0.6, 20, 10, 100], [0.8, 30, 15, 150]]


In [1370]:
from sklearn.model_selection import KFold
kf = KFold(n_splits=3) 
itr =0
idx = 0
final_acc = 0
for itr in range(len(params)):
    print("Iteration ", itr)
    print("Ratio: ", params[itr][0], ", N: ", params[itr][1], ", Max Depth: ", params[itr][2], ", Candidate Splits: ", params[itr][3])
    rf = RandomForest(ratio = params[itr][0], N = params[itr][1], max_depth = params[itr][2], candidate_splits = params[itr][3])
    accu_mean = 0
    accu_t = 0
    for train_index, test_index in kf.split(data.X_train):
        X_train, X_test = data.X_train[train_index], data.X_train[test_index]
        y_train, y_test = data.y_train[train_index], data.y_train[test_index]
        rf.fit(data.X_train, data.y_train)
        y_predict = rf.predict(X_test)
        accuracy = 0
        for i in range(len(y_predict)):
            if (y_predict[i] == y_test[i]):
                accuracy = accuracy + 1
        accu_t  += accuracy*100/len(y_predict)
    accu_mean = accu_t/3
    print("Accuracy for this iteration", accu_mean)
    if  accu_mean > final_acc:
        final_acc = accu_mean
        idx = itr
        
print("Optimal params are - Ratio: ", params[idx][0], ", N: ", params[idx][1], ", Max Depth: ", params[idx][2], ", Candidate Splits: ", params[idx][3])


Iteration  0
Ratio:  0.4 , N:  10 , Max Depth:  5 , Candidate Splits:  50
Accuracy for this iteration 83.83764597865205
Iteration  1
Ratio:  0.6 , N:  20 , Max Depth:  10 , Candidate Splits:  100
Accuracy for this iteration 90.9986367386346
Iteration  2
Ratio:  0.8 , N:  30 , Max Depth:  15 , Candidate Splits:  150
Accuracy for this iteration 92.54755409428986
Optimal params are - Ratio:  0.8 , N:  30 , Max Depth:  15 , Candidate Splits:  150


**Solution:** <br/>
From the k-fold cross validation, we see that the maximum accuracy is when ratio = 0.8, N = 30, Max Depth = 15 and Candidate Splits = 150<br/>
For depth, the  accuracy does not increase significantly beyond a certain depth (10 to 15). Increasing the depth would only overfit the data, hence it is a bad idea to do so. <br/>
We can however claim that the increasing accuracies are a result of increasing the number of learners (N=25). Increasing the ratio also has some impact on the overall accuracy as higher sample size, makes the classifiers very generalizable. <br/>
Increasing the number of candidate splits has some effect on the classifier as it does more computations to find the best split vector. However, after a certain number, it would be fruitless to go on for a large iteration. <br/>
Accuracy on these parameters is reported as below.


In [1377]:

rf = RandomForest(ratio = params[idx][0], N = params[idx][1], max_depth = params[idx][2], candidate_splits = params[idx][3])
rf.fit(data.X_train, data.y_train)
print("Fitted")
predictions = rf.predict(data.X_valid)
print("Predicted")
accuracy = 0
for i in range(len(data.y_valid)):
    if (predictions[i] == data.y_valid[i]):
        accuracy = accuracy + 1
print("Accuracy on validation set", accuracy*100/len(predictions)) 


Accuracy on validation set 92.23901466218345
